Remotion/Core/Tools/ToolRunner.cs

```diff
@@ -0,0 +1,70 @@
+﻿using System;
+using System.Configuration;
+using System.IO;
+using System.Reflection;
+using JetBrains.Annotations;
+using Remotion.Utilities;
+
+#if NET
+namespace Remotion.Tools
+{
+  /// <summary>
+  /// Base class for executing code that should run in a different base directory and or use a different config file.
+  /// </summary>
+  public abstract class ToolRunner
+  {
+    private readonly string _baseDirectory;
+    [CanBeNull]
+    private readonly string _configFile;
+
+    public ToolRunner (string baseDirectory, [CanBeNull] string configFile)
+    {
+      ArgumentUtility.CheckNotNull("baseDirectory", baseDirectory);
+
+      _baseDirectory = baseDirectory;
+      _configFile = configFile;
+    }
+
+    public void Run ()
+    {
+      var baseDirectory = Path.GetFullPath(_baseDirectory);
+      AppDomain.CurrentDomain.SetData("APP_CONTEXT_BASE_DIRECTORY", baseDirectory);
+
+      if (!string.IsNullOrEmpty(_configFile))
+      {
+        var configFile = Path.GetFullPath(_configFile);
+        if (!File.Exists(configFile))
+        {
+          throw new FileNotFoundException(
+              string.Format(
+                  "The configuration file supplied by the 'config' parameter was not found.\r\nFile: {0}",
+                  configFile),
+              configFile);
+        }
+
+        var configurationState = (int)typeof(ConfigurationManager).GetField("s_initState", BindingFlags.NonPublic | BindingFlags.Static)!.GetValue(null)!;
```

Comment:
This needs an `Assertion.IsNotNull` for the field-object. Reason: if this ever gets renamed, we fail with a null-ref-exception and not knowing what's going on is really annyong. 

Is the configuration dependency just for the assertion that ConfigurationManager wasn't initialized? If so, we could make this more "clever" by getting the ConfigurationManager assembly from the list of loaded assemblies in the AppDomain and do nothing, if it's not loaded.

I think we need a unit test for this check (i.e. testing that the reflection calls work).  If we extract the code into a ConfigurationManagerHelper class, method IsConfigurationInitialized::bool in Remotion.dll (namespace Remotion.Configuration), we have the config stuff in one place and can do a unit test, no fuzz.

Reply:
No this is a check for the internal initialization of the ConfigurationManager. If `state != 0` then the initialization did already complete and setting the config file won't have any effect since it is only considered during initialization. I included it to have a guard against somebody using this in, for example, a test project where it is probable that the config is already initialized. Unit testing this is difficult for that reason, its a singleton so if any other test has accessed it, the flag is set.

Reply:
mis-cummunication: I don't want to test that the init state is correct, just that the reflection code works correctly and returns a non-null value. Given our test suite, we may get a fixed result (e.g. 0 or 1 or 2 etc) but the important part is that we get an integer and not an expection etc.